### ISSUE TITLE
t.conch.ssh.userauth.ssh_USERAUTH_REQUEST creates a Deferred and doesn't return it

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2528]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#2528|
|Type|defect|
|Created|2007-03-19 03:21:19Z|

This is _always_ a bad thing. It makes it a pain to test and to extend.

<details><summary>Searchable metadata</summary>

```
trac-id__2528 2528
type__defect defect
reporter__jml jml
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1174274479000000 1174274479000000
changetime__1174375132000000 1174375132000000
version__None None
owner__jml jml

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_1"></a>|@jml commented|
|-|-|

Fixed in return-deferred-2528.

No, I'm not going to write tests for this.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set owner to @jml|
|-|-|

This ticket and this branch are pretty terrible.  Could you at least explain what problem is fixed by this obscure change?  (This isn't to say I won't come back later and also ask for a test)

Contrary to the ticket description, it is not _always_ bad to have a function which creates a Deferred and does not return it.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_3"></a>|@jml commented|
|-|-|

Without making this obscure change it is extraordinarily difficult to unit test Conch's authentication behaviour.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_4"></a>|@jml commented|
|-|-|

Oh, the other problem is that subclasses which extend this method are unable to add callbacks or errbacks to the Deferred. They are instead compelled to copy the method, or navigate the method's implementation to find other, Deferred-returning methods called by `ssh_USERAUTH_REQUEST`. And even then, it would make it difficult to add a callback that runs after `_ebBadAuth` is finished.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_5"></a>|@jml commented|
|-|-|

Added a test.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

Thanks.

I'm sorry that the tests ended up being so much extra work.  I am tempted to suggest ways in which they could have been done with less, but they're done already, so maybe we can talk about that for the next such branch.

We should probably just bite the bullet and codify this in the coding standard, but I've been recommending the [Divmod coding standard as a guide for spacing](http://divmod.org/trac/wiki/CodingStandard) in Twisted code for a while now.  I find that consistent use of double- and triple-blank-lines helps readability.  Can you adjust the test spacing to conform to that while you're in there?

There's a ```NameError``` logged during the test run, which is a pity, since it seems like it's actually (mostly) tangential to the purpose of the test.  Please fix.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph commented|
|-|-|

All right.  Looking good.  Merge it.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_8"></a>|@jml set status to `closed`|
|-|-|

(In [19876]) Make SSHUserAuthServer.ssh_USERAUTH_REQUEST return a Deferred.

 * Author: jml
 * Reviewers: glyph, spiv
 * Fixes [#2528](https://github.com/twisted/twisted/issues/2528)

This branch makes SSHUserAuthServer much easier to test and extend. The new 
tests are probably more interesting and helpful than the change itself.
